lib/facter/util/infiniband.rb

```diff
@@ -140,4 +140,71 @@ def self.get_hca_board_id(hca)
     board_id = self.read_sysfs(sysfs_path)
     board_id
   end
+  
+  # Returns rate of InifniBand port
+  #
+  # @return [String]
+  #
+  # @api private
+  def self.get_real_port_rate(hca, port)
+    port_sysfs_path = Dir.glob(File.join('/sys/class/infiniband', hca, 'ports', port))
+    return nil if port_sysfs_path.nil?
+
+    rate_sysfs_path = File.join(port_sysfs_path, 'rate')
+    rate = self.read_sysfs(rate_sysfs_path)
+    rate[/^(\d+)\s/,1]
+  end
+
+  # Returns state of InifniBand port
+  #
+  # @return [String]
+  #
+  # @api private
+  def self.get_real_port_state(hca, port)
+    port_sysfs_path = Dir.glob(File.join('/sys/class/infiniband', hca, 'ports', port))
+    return nil if port_sysfs_path.nil?
+
+    state_sysfs_path = File.join(port_sysfs_path, 'state')
+    state = self.read_sysfs(state_sysfs_path)
+    state[/: (.*)/,1]
+  end
+
+  # Returns link layer of Infiniband port (Ethernet or InfiniBand)
+  #
+  # @return [String]
+  #
+  # @api private
+  def self.get_real_port_linklayer(hca, port)
+    port_sysfs_path = Dir.glob(File.join('/sys/class/infiniband', hca, 'ports', port)) 
+    return nil if port_sysfs_path.nil?
+    
+    linklayer_sysfs_path = File.join(port_sysfs_path, 'link_layer')
+    linklayer = self.read_sysfs(linklayer_sysfs_path)
+    linklayer
+  end
+
+  # Returns hash of net device names (ib0, p1p1) and data about each
+  #
+  # @return [Hash]
+  #
+  # @api private
+  def self.get_netdev_to_hcaport
+    netdevs = {}
+    if ! Facter::Util::Resolution.which('ibdev2netdev')
```

Comment:
What about using a glob to see what exists in `/sys/class/infiniband`? I'd only prefer that as most of the facts rely on sysfs which is handy.

```
# /opt/puppetlabs/puppet/bin/irb 
irb(main):001:0> Dir.glob(File.join('/sys/class/infiniband', '*', 'ports', '*'))
=> ["/sys/class/infiniband/mlx5_0/ports/1", "/sys/class/infiniband/mlx5_1/ports/1", "/sys/class/infiniband/mlx5_2/ports/1", "/sys/class/infiniband/mlx5_3/ports/1"]
```

Could then infer port and hca by doing calls to something like `File.dirname()`.

Reply:
In my use case at least, I'm looking for the network interface names.  I'm not sure how to get from mlx5_0/ports/1 to ib0.  I'm sure it's somewhere other than ibdev2netdev, but I wasn't sure how to find it.  If we can get it some other way, I'm all for it.  I was concerned relying on that command could be a problem.

Reply:
Ah I had missed that you were actually pulling out `ib#` names.  I don't know of a sysfs way to pull those out as I think that's handled by udev.  Those are in /sys/class/net but I don't know how to associate /sys/class/net with /sys/class/infiniband unless there is some conversion on 64-bit mac to GUID or something similar.  I think your approach is fine.